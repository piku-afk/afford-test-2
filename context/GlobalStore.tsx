import {
  Dispatch,
  FC,
  PropsWithChildren,
  useContext,
  useReducer,
  createContext,
  useEffect,
} from 'react';
import { initialState } from './initialState';
import { ActionInterface, ActionTypes, reducer } from './reducer';

interface GlobalContextType {
  state: typeof initialState;
  dispatch: Dispatch<ActionInterface>;
}

export const GlobalContext = createContext<GlobalContextType>({
  state: initialState,
  dispatch: () => initialState,
});

export const useGlobalStore = () => useContext(GlobalContext);

type GlobalStoreProps = {};

export const GlobalStore: FC<PropsWithChildren<GlobalStoreProps>> = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    const init = async () => {
      const res = await fetch('http://localhost:5000/filters');
      const data = await res.json();
      const { searchFilter } = data || {};
      const { brands, categories } = searchFilter || {};

      dispatch({ type: ActionTypes.setCategories, payload: categories });
      dispatch({ type: ActionTypes.setBrands, payload: brands });
    };
    init();
  }, []);

  const value = { state, dispatch };

  return (
    <GlobalContext.Provider value={value}>{children}</GlobalContext.Provider>
  );
};
