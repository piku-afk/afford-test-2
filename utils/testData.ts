export const mockCategories = [
  {
    ID: '1',
    name: 'Hello',
    subCategories: [
      { ID: '1', name: 'Piyush', subCategories: [] },
      { ID: '2', name: 'Mahato', subCategories: [] },
    ],
  },
  { ID: '2', name: 'World 2', subCategories: [] },
];

export const mockProducts = [
  {
    id: '1',
    title: 'product 1',
    description: 'some description',
    price: '',
    category: '',
    image: 'http://placeimg.com',
    quantity: 1,
    rating: { rate: '', count: '' },
  },
];

export const mockBrands = [
  { ID: '1', name: 'hello' },
  { ID: '2', name: 'world' },
];
