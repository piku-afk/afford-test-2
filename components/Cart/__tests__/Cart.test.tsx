import { cleanup, render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Cart } from '../index';
import * as GlobalContext from '@/context/GlobalStore';
import { mockProducts } from '@/utils/testData';
import { ActionTypes } from '@/context/reducer';

const user = userEvent.setup();

const MockCart = () => (
  <Cart id='hello-world' anchorElement={{} as Element} handleClose={jest.fn} />
);
const dispatch = jest.fn();

describe('Cart', () => {
  beforeEach(() => {
    jest.spyOn(GlobalContext, 'useGlobalStore').mockReturnValue({
      state: { categories: [], brands: [], cart: mockProducts },
      dispatch,
    });
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<MockCart />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders different elements', () => {
    render(<MockCart />);

    const cartPopup = screen.getByRole('presentation');
    const heading = screen.getByRole('heading', { name: /shopping cart/i });
    const closeButton = screen.getByRole('button', { name: /close-cart/i });
    const continueShopping = screen.getByText(/continue shopping/i);
    const checkoutButton = screen.getByRole('button', {
      name: /go to checkout/i,
    });

    expect(cartPopup).toHaveAttribute('id', 'hello-world');
    expect(heading).toBeInTheDocument();
    expect(closeButton).toBeInTheDocument();
    expect(continueShopping).toBeInTheDocument();
    expect(checkoutButton).toBeInTheDocument();
  });

  it('renders empty cart initially', () => {
    jest.spyOn(GlobalContext, 'useGlobalStore').mockReturnValue({
      state: { categories: [], brands: [], cart: [] },
      dispatch,
    });

    render(<MockCart />);

    const emptyCart = screen.getByText(/empty cart/i);
    expect(emptyCart).toBeInTheDocument();
  });

  it('render products in cart', async () => {
    render(<MockCart />);

    const emptyCart = screen.queryByText(/empty cart/i);
    const products = screen.getAllByTestId('product-card');

    expect(emptyCart).toBe(null);
    expect(products).toHaveLength(1);

    await user.click(
      within(products[0]).getByRole('button', { name: /remove/i })
    );

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(
      expect.objectContaining({
        type: ActionTypes.removeFromCart,
        payload: mockProducts[0],
      })
    );
  });
});

export {};
