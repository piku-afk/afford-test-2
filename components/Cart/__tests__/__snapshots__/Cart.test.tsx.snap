// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart matches snapshot 1`] = `<DocumentFragment />`;
